corewars-sim/src/core.rs

```diff
@@ -130,32 +126,64 @@ impl Core {
             self.instructions[i] = instruction.clone();
         }
 
-        self.entry_point
-            .set_value(warrior.program.origin.unwrap_or(0) as _);
-        self.program_counter = self.entry_point;
+        // TODO: Maybe some kinda increasing counter for warrior names
+        let warrior_name = warrior
+            .metadata
+            .name
+            .clone()
+            .unwrap_or_else(|| String::from("Warrior0"));
+
+        self.process_queue.push(
+            warrior_name,
+            self.offset(warrior.program.origin.unwrap_or(0) as i32),
+        );
 
         Ok(())
     }
 
     /// Run a single cycle of simulation. This will continue to execute even
     /// after MAXCYCLES has been reached
-    pub fn step(&mut self) -> Result<(), ExecutionError> {
+    pub fn step(&mut self) -> Result<(), process::Error> {
         self.steps_taken += 1;
-        let result = opcode::execute(self)?;
 
-        // If the opcode affected the program counter, avoid incrementing it an extra time
-        if let Some(offset) = result.program_counter_offset {
-            self.program_counter += offset;
-        } else {
-            self.program_counter += 1;
-        }
+        let result = opcode::execute(self);
+        let current_process = self.process_queue.pop()?;
```

Comment:
These were swapped since it fixed a bug in `wrap_program_counter_on_overflow`... but actually, the real fix is to swap back and pass through the offset into `opcode::execute`, I think.